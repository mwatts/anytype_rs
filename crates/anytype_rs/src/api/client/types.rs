//! Type management module
//!
//! Handles type management operations such as creating, updating, and deleting object types.

use super::AnytypeClient;
use crate::{api::types::Icon, error::Result, types::Pagination};
use serde::{Deserialize, Serialize};
use tracing::{debug, info};

/// Property format for type creation
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "snake_case")]
pub enum PropertyFormat {
    Text,
    Number,
    Select,
    MultiSelect,
    Date,
    Files,
    Checkbox,
    Url,
    Email,
    Phone,
    Objects,
}

/// Layout type for object types
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Layout {
    Basic,
    Profile,
    Action,
    Note,
    Bookmark,
    Set,
    Collection,
    Participant,
}

/// Property for type creation
#[derive(Debug, Serialize, Deserialize)]
pub struct CreateTypeProperty {
    pub format: PropertyFormat,
    pub key: String,
    pub name: String,
}

/// Request to create a new type
#[derive(Debug, Serialize)]
pub struct CreateTypeRequest {
    pub icon: Icon,
    pub key: String,
    pub layout: Layout,
    pub name: String,
    pub plural_name: String,
    pub properties: Vec<CreateTypeProperty>,
}

/// Response when creating a type
#[derive(Debug, Deserialize)]
pub struct CreateTypeResponse {
    #[serde(rename = "type")]
    pub type_data: Type,
}

/// Response when getting a type
#[derive(Debug, Deserialize)]
pub struct GetTypeResponse {
    #[serde(rename = "type")]
    pub type_data: Type,
}

/// Property information for types
#[derive(Debug, Deserialize, Serialize)]
pub struct TypeProperty {
    pub format: String,
    pub id: String,
    pub key: String,
    pub name: String,
    pub object: String,
}

/// Type information
#[derive(Debug, Deserialize, Serialize)]
pub struct Type {
    pub archived: Option<bool>,
    pub icon: Icon,
    pub id: String,
    pub key: String,
    pub layout: Option<String>,
    pub name: String,
    pub object: String,
    pub plural_name: Option<String>,
    pub properties: Vec<TypeProperty>,
}

/// Response for listing types
#[derive(Debug, Deserialize)]
pub struct ListTypesResponse {
    pub data: Vec<Type>,
    pub pagination: Pagination,
}

/// Request to update an existing type
#[derive(Debug, Serialize)]
pub struct UpdateTypeRequest {
    pub icon: Icon,
    pub key: String,
    pub layout: Layout,
    pub name: String,
    pub plural_name: String,
    pub properties: Vec<CreateTypeProperty>,
}

/// Response when updating a type
#[derive(Debug, Deserialize)]
pub struct UpdateTypeResponse {
    #[serde(rename = "type")]
    pub type_data: Type,
}

/// Response when deleting a type
#[derive(Debug, Deserialize)]
pub struct DeleteTypeResponse {
    #[serde(rename = "type")]
    pub type_data: Type,
}

impl AnytypeClient {
    /// List types in a space
    pub async fn list_types(&self, space_id: &str) -> Result<Vec<Type>> {
        let response: ListTypesResponse = self.get(&format!("/v1/spaces/{space_id}/types")).await?;
        Ok(response.data)
    }

    /// List types in a space with pagination information
    pub async fn list_types_with_pagination(&self, space_id: &str) -> Result<ListTypesResponse> {
        info!("Listing types in space: {}", space_id);
        debug!("GET /v1/spaces/{}/types", space_id);

        self.get(&format!("/v1/spaces/{space_id}/types")).await
    }

    /// Create a new type in a space
    pub async fn create_type(
        &self,
        space_id: &str,
        request: CreateTypeRequest,
    ) -> Result<CreateTypeResponse> {
        info!("Creating type '{}' in space: {}", request.name, space_id);
        debug!("Request: {:?}", request);
        debug!("Request JSON: {}", serde_json::to_string_pretty(&request)?);

        self.post(&format!("/v1/spaces/{space_id}/types"), &request)
            .await
    }

    /// Get a specific type by ID
    pub async fn get_type(&self, space_id: &str, type_id: &str) -> Result<Type> {
        info!("Getting type '{}' from space: {}", type_id, space_id);
        debug!("GET /v1/spaces/{}/types/{}", space_id, type_id);

        let response: GetTypeResponse = self
            .get(&format!("/v1/spaces/{space_id}/types/{type_id}"))
            .await?;
        Ok(response.type_data)
    }

    /// Update an existing type in a space
    pub async fn update_type(
        &self,
        space_id: &str,
        type_id: &str,
        request: UpdateTypeRequest,
    ) -> Result<UpdateTypeResponse> {
        info!("Updating type '{}' in space: {}", type_id, space_id);
        debug!("Request: {:?}", request);
        debug!("Request JSON: {}", serde_json::to_string_pretty(&request)?);

        self.patch(&format!("/v1/spaces/{space_id}/types/{type_id}"), &request)
            .await
    }

    /// Delete (archive) a type in a space
    pub async fn delete_type(&self, space_id: &str, type_id: &str) -> Result<DeleteTypeResponse> {
        info!(
            "Deleting (archiving) type '{}' in space: {}",
            type_id, space_id
        );
        debug!("DELETE /v1/spaces/{}/types/{}", space_id, type_id);

        self.delete(&format!("/v1/spaces/{space_id}/types/{type_id}"))
            .await
    }
}
