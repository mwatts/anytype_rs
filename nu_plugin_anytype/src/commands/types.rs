use crate::{commands::common::get_space_id, value::AnytypeValue, AnytypePlugin};
use nu_plugin::{EngineInterface, EvaluatedCall, PluginCommand};
use nu_protocol::{Category, LabeledError, PipelineData, Signature, SyntaxShape, Value};

/// Command: anytype type list
pub struct TypeList;

impl PluginCommand for TypeList {
    type Plugin = AnytypePlugin;

    fn name(&self) -> &str {
        "anytype type list"
    }

    fn description(&self) -> &str {
        "List all types in a space"
    }

    fn signature(&self) -> Signature {
        Signature::build(self.name())
            .named(
                "space",
                SyntaxShape::String,
                "Name of the space (can also accept Space from pipeline)",
                Some('s'),
            )
            .input_output_types(vec![
                (
                    nu_protocol::Type::Nothing,
                    nu_protocol::Type::List(Box::new(nu_protocol::Type::Custom("AnytypeValue".into()))),
                ),
                (
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                    nu_protocol::Type::List(Box::new(nu_protocol::Type::Custom("AnytypeValue".into()))),
                ),
            ])
            .category(Category::Custom("anytype".into()))
    }

    fn run(
        &self,
        plugin: &Self::Plugin,
        _engine: &EngineInterface,
        call: &EvaluatedCall,
        input: PipelineData,
    ) -> Result<PipelineData, LabeledError> {
        let span = call.head;
        let input = input.into_value(span)?;

        // Get space_id from multiple sources
        let space_id = get_space_id(plugin, call, &input, span)?;

        // Get client
        let client = plugin.client().map_err(|e| {
            LabeledError::new(format!("Failed to get client: {}", e))
                .with_label("Authentication required", span)
        })?;

        // List types from API
        let types = plugin
            .run_async(client.list_types(&space_id))
            .map_err(|e| LabeledError::new(format!("Failed to list types: {}", e)))?;

        // Convert to AnytypeValue::Type with space_id context
        let values: Vec<Value> = types
            .into_iter()
            .map(|type_data| {
                // Use From<(Type, String)> to convert with space_id context
                let anytype_value: AnytypeValue = (type_data, space_id.clone()).into();
                Value::custom(Box::new(anytype_value), span)
            })
            .collect();

        Ok(PipelineData::Value(Value::list(values, span), None))
    }
}

/// Command: anytype type get
pub struct TypeGet;

impl PluginCommand for TypeGet {
    type Plugin = AnytypePlugin;

    fn name(&self) -> &str {
        "anytype type get"
    }

    fn description(&self) -> &str {
        "Get a type by name"
    }

    fn signature(&self) -> Signature {
        Signature::build(self.name())
            .required("name", SyntaxShape::String, "Name of the type")
            .named(
                "space",
                SyntaxShape::String,
                "Name of the space (can also accept Space from pipeline)",
                Some('s'),
            )
            .input_output_types(vec![
                (
                    nu_protocol::Type::Nothing,
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                ),
                (
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                ),
            ])
            .category(Category::Custom("anytype".into()))
    }

    fn run(
        &self,
        plugin: &Self::Plugin,
        _engine: &EngineInterface,
        call: &EvaluatedCall,
        input: PipelineData,
    ) -> Result<PipelineData, LabeledError> {
        let span = call.head;
        let input = input.into_value(span)?;

        // Get type name from arguments
        let name: String = call.req(0)?;

        // Get space_id from multiple sources
        let space_id = get_space_id(plugin, call, &input, span)?;

        // Get resolver
        let resolver = plugin.resolver().map_err(|e| {
            LabeledError::new(format!("Failed to get resolver: {}", e))
                .with_label("Authentication required", span)
        })?;

        // Resolve type name to ID within the space
        let type_id = plugin
            .run_async(resolver.resolve_type(&space_id, &name))
            .map_err(|e| {
                LabeledError::new(format!(
                    "Failed to resolve type '{}' in space '{}': {}",
                    name, space_id, e
                ))
            })?;

        // Get client
        let client = plugin.client().map_err(|e| {
            LabeledError::new(format!("Failed to get client: {}", e))
                .with_label("Authentication required", span)
        })?;

        // Fetch type details
        let type_data = plugin
            .run_async(client.get_type(&space_id, &type_id))
            .map_err(|e| LabeledError::new(format!("Failed to get type: {}", e)))?;

        // Convert to AnytypeValue::Type with space_id context
        let anytype_value: AnytypeValue = (type_data, space_id).into();
        Ok(PipelineData::Value(Value::custom(Box::new(anytype_value), span), None))
    }
}

