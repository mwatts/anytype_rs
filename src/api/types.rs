use serde::{Deserialize, Serialize};

/// Color for tags and icons
#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, strum::Display)]
#[serde(rename_all = "lowercase")]
pub enum Color {
    Grey,
    Yellow,
    Orange,
    Red,
    Pink,
    Purple,
    Blue,
    Ice,
    Teal,
    Lime,
}

/// Generic API error response
#[derive(Debug, Deserialize)]
pub struct ApiErrorResponse {
    pub code: String,
    pub message: String,
    pub object: String,
    pub status: u32,
}

/// Pagination information
#[derive(Debug, Deserialize, Serialize)]
pub struct Pagination {
    pub has_more: bool,
    pub limit: usize,
    pub offset: usize,
    pub total: usize,
}

/// Icon format type
#[derive(Debug, Deserialize, Serialize, Clone)]
pub enum IconFormat {
    Emoji,
    File,
    Icon,
}

/// Icon enum that can be one of three types
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "format")]
pub enum Icon {
    #[serde(rename = "emoji")]
    Emoji { emoji: String },
    #[serde(rename = "file")]
    File { file: String },
    #[serde(rename = "icon")]
    Icon { color: Color, name: String },
}

/// Property information for types
#[derive(Debug, Deserialize, Serialize)]
pub struct TypeProperty {
    pub format: Format,
    pub id: String,
    pub key: String,
    pub name: String,
    pub object: String,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "snake_case")]
pub enum Format {
    Text,
    Number,
    Select,
    MultiSelect,
    Date,
    Files,
    Checkbox,
    Url,
    Email,
    Phone,
    Objects,
}

/// Type information
#[derive(Debug, Deserialize, Serialize)]
pub struct Type {
    pub archived: bool,
    pub icon: Option<Icon>,
    pub id: String,
    pub key: String,
    pub layout: Layout,
    pub name: String,
    pub object: String,
    pub plural_name: String,
    pub properties: Vec<TypeProperty>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "lowercase")]
pub enum Layout {
    Basic,
    Profile,
    Action,
    Note,
    Bookmark,
    Set,
    Collection,
    Participant,
}
