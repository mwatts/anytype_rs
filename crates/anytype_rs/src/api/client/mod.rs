//! Anytype API client modules
//!
//! This module is organized to match the official API reference structure.

use crate::{error::Result, types::ApiErrorResponse};
use reqwest::{Client, Method, RequestBuilder};
use serde::{Serialize, de::DeserializeOwned};
use tracing::{debug, error};

// Include all module implementations
pub mod auth;
pub mod lists;
pub mod members;
pub mod objects;
pub mod properties;
pub mod search;
pub mod spaces;
pub mod tags;
pub mod templates;
pub mod types;

const DEFAULT_BASE_URL: &str = "http://localhost:31009";
const ANYTYPE_API_HEADER: &str = "Anytype-Version";
// TODO: Better support multiple API versions
const ANYTYPE_API_VERSION: &str = "2025-05-20";

/// Configuration for the Anytype client
#[derive(Debug, Clone)]
pub struct ClientConfig {
    pub base_url: String,
    pub timeout_seconds: u64,
    pub app_name: String,
}

impl Default for ClientConfig {
    fn default() -> Self {
        Self {
            base_url: DEFAULT_BASE_URL.to_string(),
            timeout_seconds: 30,
            app_name: "anytype_rs".to_string(),
        }
    }
}

/// Main client for interacting with the Anytype API
#[derive(Debug)]
pub struct AnytypeClient {
    pub(crate) http_client: Client,
    pub(crate) config: ClientConfig,
    pub(crate) api_key: Option<String>,
}

impl AnytypeClient {
    /// Create a new client with default configuration
    pub fn new() -> Result<Self> {
        Self::with_config(ClientConfig::default())
    }

    /// Create a new client with custom configuration
    pub fn with_config(config: ClientConfig) -> Result<Self> {
        let http_client = Client::builder()
            .timeout(std::time::Duration::from_secs(config.timeout_seconds))
            .build()?;

        Ok(Self {
            http_client,
            config,
            api_key: None,
        })
    }

    /// Set the API key for authenticated requests
    pub fn set_api_key(&mut self, api_key: String) {
        self.api_key = Some(api_key);
    }

    /// Get the current API key
    pub fn api_key(&self) -> Option<&str> {
        self.api_key.as_deref()
    }

    /// Make an authenticated GET request
    pub(crate) async fn get<T: DeserializeOwned>(&self, path: &str) -> Result<T> {
        let url = format!("{}{}", self.config.base_url, path);
        debug!("GET {}", url);

        let response = self
            .authenticated_request(Method::GET, &url)?
            .send()
            .await?;

        self.handle_response(response).await
    }

    /// Make an authenticated POST request with JSON body
    pub(crate) async fn post<T: DeserializeOwned, B: Serialize>(
        &self,
        path: &str,
        body: &B,
    ) -> Result<T> {
        let url = format!("{}{}", self.config.base_url, path);
        debug!("POST {}", url);

        let response = self
            .authenticated_request(Method::POST, &url)?
            .json(body)
            .send()
            .await?;

        self.handle_response(response).await
    }

    /// Make an authenticated PATCH request with JSON body
    pub(crate) async fn patch<T: DeserializeOwned, B: Serialize>(
        &self,
        path: &str,
        body: &B,
    ) -> Result<T> {
        let url = format!("{}{}", self.config.base_url, path);
        debug!("PATCH {}", url);

        let response = self
            .authenticated_request(Method::PATCH, &url)?
            .json(body)
            .send()
            .await?;

        self.handle_response(response).await
    }

    /// Make an authenticated DELETE request
    pub(crate) async fn delete<T: DeserializeOwned>(&self, path: &str) -> Result<T> {
        let url = format!("{}{}", self.config.base_url, path);
        debug!("DELETE {}", url);

        let response = self
            .authenticated_request(Method::DELETE, &url)?
            .send()
            .await?;

        self.handle_response(response).await
    }

    /// Make an unauthenticated POST request (for auth endpoints)
    pub(crate) async fn post_unauthenticated<T: DeserializeOwned, B: Serialize>(
        &self,
        path: &str,
        body: &B,
    ) -> Result<T> {
        let url = format!("{}{}", self.config.base_url, path);
        debug!("POST {} (unauthenticated)", url);

        let response = self
            .http_client
            .post(&url)
            .header(ANYTYPE_API_HEADER, ANYTYPE_API_VERSION)
            .json(body)
            .send()
            .await?;

        self.handle_response(response).await
    }

    /// Create an authenticated request builder (internal helper)
    fn authenticated_request(&self, method: Method, url: &str) -> Result<RequestBuilder> {
        let api_key = self
            .api_key
            .as_ref()
            .ok_or_else(|| crate::error::AnytypeError::Auth {
                message: "API key not set. Call set_api_key() first.".to_string(),
            })?;

        let builder = match method {
            Method::GET => self.http_client.get(url),
            Method::POST => self.http_client.post(url),
            Method::PATCH => self.http_client.patch(url),
            Method::DELETE => self.http_client.delete(url),
            _ => {
                return Err(crate::error::AnytypeError::Api {
                    message: format!("Unsupported HTTP method: {method}"),
                });
            }
        };

        Ok(builder
            .header(ANYTYPE_API_HEADER, ANYTYPE_API_VERSION)
            .bearer_auth(api_key))
    }

    /// Handle HTTP response and deserialize JSON
    pub(crate) async fn handle_response<T: DeserializeOwned>(
        &self,
        response: reqwest::Response,
    ) -> Result<T> {
        let status = response.status();
        let url = response.url().clone();

        debug!("Response status: {} for {}", status, url);

        if status.is_success() {
            // Get the response text first for debugging
            let response_text =
                response
                    .text()
                    .await
                    .map_err(|e| crate::error::AnytypeError::InvalidResponse {
                        message: format!("Failed to read response body: {e}"),
                    })?;

            debug!("Response body: {response_text}");

            let response = serde_json::from_str::<T>(&response_text);

            match response {
                Ok(data) => Ok(data),
                Err(e) => {
                    error!("Failed to deserialize response: {}", e);
                    error!("Expected type: {}", std::any::type_name::<T>());
                    error!("Response body was: {}", response_text);
                    Err(crate::error::AnytypeError::InvalidResponse {
                        message: format!(
                            "Failed to parse JSON response: {}. Expected type: {}",
                            e,
                            std::any::type_name::<T>()
                        ),
                    })
                }
            }
        } else {
            let response = response.json::<ApiErrorResponse>().await;
            error!("API error {}", status);

            match response {
                Ok(error) => {
                    let message = error.message;

                    if status == 401 || status == 403 {
                        Err(crate::error::AnytypeError::Auth { message })
                    } else {
                        Err(crate::error::AnytypeError::Api { message })
                    }
                }
                Err(e) => Err(crate::error::AnytypeError::Api {
                    message: format!("HTTP {status} - {e}"),
                }),
            }
        }
    }
}
