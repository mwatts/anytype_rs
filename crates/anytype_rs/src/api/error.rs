use thiserror::Error;

#[derive(Error, Debug)]
pub enum AnytypeError {
    #[error("HTTP request failed: {source}")]
    Http {
        #[from]
        source: reqwest::Error,
    },

    #[error("Authentication failed: {message}")]
    Auth { message: String },

    #[error("API error: {message}")]
    Api { message: String },

    #[error("Serialization error: {source}")]
    Serialization {
        #[from]
        source: serde_json::Error,
    },

    #[error("Invalid response: {message}")]
    InvalidResponse { message: String },
}

pub type Result<T> = std::result::Result<T, AnytypeError>;
