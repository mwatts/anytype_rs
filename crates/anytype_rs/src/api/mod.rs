//! # Anytype API Client
//!
//! A Rust library for interacting with your local Anytype application API.
//!
//! ## Features
//!
//! - Authentication via challenge-response mechanism with local Anytype app
//! - JWT Bearer token support
//! - Full CRUD operations for spaces and objects
//! - Search functionality
//! - Template, type, property, and tag management
//! - Async/await support with tokio
//! - Comprehensive error handling

pub mod client;
pub mod error;
pub mod types;

pub use client::{AnytypeClient, ClientConfig};
pub use error::{AnytypeError, Result};
pub use types::*;

// Re-export types from client modules for convenience
pub use client::auth::{
    CreateApiKeyRequest, CreateApiKeyResponse, CreateChallengeRequest, CreateChallengeResponse,
};
pub use client::lists::{
    AddListObjectsRequest, AddListObjectsResponse, GetListObjectsResponse, GetListViewsResponse,
    ListObject, ListObjectType, ListViewData, ListViewFilter, ListViewSort, ObjectTypeProperty,
    RemoveListObjectsResponse,
};
pub use client::members::{
    GetMemberResponse, ListMembersResponse, Member, MemberRole, MemberStatus,
};
pub use client::objects::{
    CreateObjectRequest, CreateObjectResponse, DeleteObjectResponse, ListObjectsResponse, Object,
    UpdateObjectRequest, UpdateObjectResponse,
};
pub use client::properties::{
    CreatePropertyRequest, CreatePropertyResponse, DeletePropertyResponse, GetPropertyResponse,
    ListPropertiesResponse, Property, UpdatePropertyRequest, UpdatePropertyResponse,
};
pub use client::search::{
    SearchObject, SearchRequest, SearchResponse, SearchSpaceRequest, Sort, SortDirection,
    SortProperty,
};
pub use client::spaces::{
    CreateSpaceRequest, CreateSpaceResponse, ListSpacesResponse, Space, UpdateSpaceRequest,
    UpdateSpaceResponse,
};
pub use client::tags::{
    CreateTagRequest, CreateTagResponse, DeleteTagResponse, GetTagResponse, ListTagsResponse, Tag,
    UpdateTagRequest, UpdateTagResponse,
};
pub use client::templates::{GetTemplateResponse, ListTemplatesResponse, ObjectType, Template};
pub use client::types::{
    CreateTypeProperty, CreateTypeRequest, CreateTypeResponse, DeleteTypeResponse, GetTypeResponse,
    Layout, ListTypesResponse, PropertyFormat, Type, TypeProperty, UpdateTypeRequest,
    UpdateTypeResponse,
};
pub use types::{Icon, IconFormat};
