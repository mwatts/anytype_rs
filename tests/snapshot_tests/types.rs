//! Snapshot tests for core API types

use anytype_rs::api::types::{Color, Format, Icon, IconFormat, Layout, Pagination, Type, TypeProperty};

#[test]
fn test_color_serialization() {
    insta::assert_json_snapshot!("color_grey", Color::Grey);
    insta::assert_json_snapshot!("color_yellow", Color::Yellow);
    insta::assert_json_snapshot!("color_orange", Color::Orange);
    insta::assert_json_snapshot!("color_red", Color::Red);
    insta::assert_json_snapshot!("color_pink", Color::Pink);
    insta::assert_json_snapshot!("color_purple", Color::Purple);
    insta::assert_json_snapshot!("color_blue", Color::Blue);
    insta::assert_json_snapshot!("color_ice", Color::Ice);
    insta::assert_json_snapshot!("color_teal", Color::Teal);
    insta::assert_json_snapshot!("color_lime", Color::Lime);
}

#[test]
fn test_format_serialization() {
    insta::assert_json_snapshot!("format_text", Format::Text);
    insta::assert_json_snapshot!("format_number", Format::Number);
    insta::assert_json_snapshot!("format_select", Format::Select);
    insta::assert_json_snapshot!("format_multi_select", Format::MultiSelect);
    insta::assert_json_snapshot!("format_date", Format::Date);
    insta::assert_json_snapshot!("format_files", Format::Files);
    insta::assert_json_snapshot!("format_checkbox", Format::Checkbox);
    insta::assert_json_snapshot!("format_url", Format::Url);
    insta::assert_json_snapshot!("format_email", Format::Email);
    insta::assert_json_snapshot!("format_phone", Format::Phone);
    insta::assert_json_snapshot!("format_objects", Format::Objects);
}

#[test]
fn test_layout_serialization() {
    insta::assert_json_snapshot!("layout_basic", Layout::Basic);
    insta::assert_json_snapshot!("layout_profile", Layout::Profile);
    insta::assert_json_snapshot!("layout_action", Layout::Action);
    insta::assert_json_snapshot!("layout_note", Layout::Note);
    insta::assert_json_snapshot!("layout_bookmark", Layout::Bookmark);
    insta::assert_json_snapshot!("layout_set", Layout::Set);
    insta::assert_json_snapshot!("layout_collection", Layout::Collection);
    insta::assert_json_snapshot!("layout_participant", Layout::Participant);
}

#[test]
fn test_icon_format_serialization() {
    insta::assert_json_snapshot!("icon_format_emoji", IconFormat::Emoji);
    insta::assert_json_snapshot!("icon_format_file", IconFormat::File);
    insta::assert_json_snapshot!("icon_format_icon", IconFormat::Icon);
}

#[test]
fn test_icon_variants_serialization() {
    let emoji_icon = Icon::Emoji {
        emoji: "üòÄ".to_string(),
    };
    insta::assert_json_snapshot!("icon_emoji", emoji_icon);

    let file_icon = Icon::File {
        file: "path/to/icon.png".to_string(),
    };
    insta::assert_json_snapshot!("icon_file", file_icon);

    let color_icon = Icon::Icon {
        color: Color::Blue,
        name: "star".to_string(),
    };
    insta::assert_json_snapshot!("icon_color", color_icon);
}

#[test]
fn test_pagination_serialization() {
    let pagination = Pagination {
        has_more: true,
        limit: 50,
        offset: 100,
        total: 250,
    };
    insta::assert_json_snapshot!("pagination_with_more", pagination);

    let pagination_no_more = Pagination {
        has_more: false,
        limit: 50,
        offset: 200,
        total: 250,
    };
    insta::assert_json_snapshot!("pagination_no_more", pagination_no_more);
}

#[test]
fn test_type_property_serialization() {
    let type_property = TypeProperty {
        format: Format::Text,
        id: "prop123".to_string(),
        key: "title".to_string(),
        name: "Title".to_string(),
        object: "property".to_string(),
    };
    insta::assert_json_snapshot!("type_property", type_property);
}

#[test]
fn test_type_serialization() {
    let type_obj = Type {
        archived: false,
        icon: Some(Icon::Emoji {
            emoji: "üìù".to_string(),
        }),
        id: "type123".to_string(),
        key: "note".to_string(),
        layout: Layout::Note,
        name: "Note".to_string(),
        object: "type".to_string(),
        plural_name: "Notes".to_string(),
        properties: vec![
            TypeProperty {
                format: Format::Text,
                id: "prop1".to_string(),
                key: "title".to_string(),
                name: "Title".to_string(),
                object: "property".to_string(),
            },
            TypeProperty {
                format: Format::Date,
                id: "prop2".to_string(),
                key: "created".to_string(),
                name: "Created".to_string(),
                object: "property".to_string(),
            },
        ],
    };
    insta::assert_json_snapshot!("type_with_properties", type_obj);
}
