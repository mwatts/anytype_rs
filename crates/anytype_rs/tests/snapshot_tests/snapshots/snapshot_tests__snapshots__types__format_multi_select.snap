---
source: tests/snapshot_tests/types.rs
expression: "Format::MultiSelect"
---
"multi_select"
